curvefs/src/mds/topology/topology_metric.cpp

```diff
@@ -0,0 +1,204 @@
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curvefs
+ * Created Date: 2021/11/1
+ * Author: chenwei
+ */
+
+#include "curvefs/src/mds/topology/topology_metric.h"
+
+#include <memory>
+#include <set>
+#include <list>
+
+namespace curvefs {
+namespace mds {
+namespace topology {
+
+std::map<PoolIdType, PoolMetricPtr> gPoolMetrics;
+std::map<MetaServerIdType, MetaServerMetricPtr> gMetaServerMetrics;
+
+void TopologyMetricService::UpdateTopologyMetrics() {
+    // process metaserver
+    std::vector<MetaServerIdType> metaservers = topo_->GetMetaServerInCluster(
+        [](const MetaServer &ms) { return true; });
+
+    for (auto msId : metaservers) {
+        auto it = gMetaServerMetrics.find(msId);
+        if (it == gMetaServerMetrics.end()) {
+            MetaServerMetricPtr cptr(new MetaServerMetric(msId));
+            it = gMetaServerMetrics.emplace(msId, std::move(cptr)).first;
+        }
+        MetaServer ms;
+        if (topo_->GetMetaServer(msId, &ms)) {
+            it->second->diskCapacity.set_value(
+                ms.GetMetaServerSpace().GetDiskCapacity());
+            it->second->diskUsed.set_value(
+                ms.GetMetaServerSpace().GetDiskUsed());
+            it->second->memoryUsed.set_value(
+                ms.GetMetaServerSpace().GetMemoryUsed());
+        }
+    }
+
+    // process pool
+    std::vector<PoolIdType> pools = topo_->GetPoolInCluster(
+        [](const Pool &pool) { return pool.GetPoolAvaliableFlag(); });
+    for (auto pid : pools) {
+        Pool pool;
+        if (!topo_->GetPool(pid, &pool)) {
+            continue;
+        }
+        std::string poolName = pool.GetName();
+
+        std::vector<CopySetInfo> copysets = topo_->GetCopySetInfosInPool(pid);
+
+        std::map<MetaServerIdType, MetaServerMetricInfo> metaServerMetricInfo;
+        CalcMetaServerMetrics(copysets, &metaServerMetricInfo);
+
+        auto it = gPoolMetrics.find(pid);
+        if (it == gPoolMetrics.end()) {
+            PoolMetricPtr lptr(new PoolMetric(poolName));
+            it = gPoolMetrics.emplace(pid, std::move(lptr)).first;
+        }
+
+        it->second->metaServerNum.set_value(metaServerMetricInfo.size());
+        it->second->copysetNum.set_value(copysets.size());
+
+        uint64_t totalInodeNum = 0;
+        uint64_t totalDentryNum = 0;
+        std::list<Partition> partitions = topo_->GetPartitionInfosInPool(pid);
+        for (auto pit = partitions.begin(); pit != partitions.end(); ++pit) {
+            totalInodeNum += pit->GetInodeNum();
+            totalDentryNum += pit->GetDentryNum();
+        }
+        it->second->inodeNum.set_value(totalInodeNum);
+        it->second->dentryNum.set_value(totalDentryNum);
+        it->second->partitionNum.set_value(partitions.size());
+
+
+        uint64_t totalDiskCapacity = 0;
+        uint64_t totalDiskUsed = 0;
+
+        // process the metric of metaserver.
+        for (auto cm : metaServerMetricInfo) {
```

Comment:
for (const auto &cm : metaServerMetricInfo)

Reply:
done